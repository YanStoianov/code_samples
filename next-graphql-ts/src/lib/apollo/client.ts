import { ApolloClient, InMemoryCache, HttpLink, ApolloLink, concat } from "@apollo/client";
import { CookieValueTypes, getCookie } from "cookies-next";
import { userConstants } from "@/store/user/constants";
import cacheOptions from "./cacheOptions";

interface ApolloClientProps {
  token?: CookieValueTypes;
}

const httpLink = new HttpLink({ uri: process.env.NEXT_PUBLIC_APOLLO_URI, includeExtensions: true });
const authToken = getCookie(userConstants.ACCESS_TOKEN);

const authMiddleware = ({ token }: ApolloClientProps) => {
  return new ApolloLink((operation, forward) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    operation.setContext(({ headers }: Record<string, any>) => {
      return {
        headers: {
          authorization: token ? `Bearer ${token}` : "Bearer guest",
          ...headers,
        },
      };
    });
    return forward(operation).map((response) => {
      if (response.data) {
        response.data.extensions = response.extensions;
      }
      return response;
    });
  });
};

const client = new ApolloClient({
  ssrMode: true,
  link: concat(authMiddleware({ token: authToken }), httpLink),
  cache: new InMemoryCache(cacheOptions),
});

export const getApolloClient = ({ token }: ApolloClientProps) => {
  return new ApolloClient({
    ssrMode: true,
    link: concat(authMiddleware({ token }), httpLink),
    cache: new InMemoryCache(cacheOptions),
  });
};

export default client;
