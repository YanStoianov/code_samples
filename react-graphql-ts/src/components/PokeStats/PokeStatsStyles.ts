import styled from 'styled-components';

export default styled.div`
  .stats-block {
    display: flex;
    justify-content: space-around;
    padding: 20px;
    color: gray;
  }
`;