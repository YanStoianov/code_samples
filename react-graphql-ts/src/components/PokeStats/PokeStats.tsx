import React, {useMemo} from "react";
import PokeStatsStyles from "./PokeStatsStyles";
import {useQuery} from "@apollo/client";
import {GET_POKEMONS_COUNT} from "../../queries/pokQueries";
import {PokemonTypes} from "../../interfaces/pokeInterfaces";

interface PokeStatsProps {
    pokemonTypes: PokemonTypes | undefined
}

const PokeStats = (props: PokeStatsProps) => {
    const {pokemonTypes} = props;
    const {data: uniquePokemonsData} = useQuery(GET_POKEMONS_COUNT, {
        variables: {
            default: true
        }
    });
    const {data: variationsPokemonData} = useQuery(GET_POKEMONS_COUNT, {
        variables: {
            default: false
        }
    });

    const stats = useMemo(() => {
        const uniquePokemonCount = uniquePokemonsData?.pokemon_v2_pokemon_aggregate?.aggregate?.count || 0;
        const duplicationPokemonCount = variationsPokemonData?.pokemon_v2_pokemon_aggregate?.aggregate?.count || 0;
        const typesCount = pokemonTypes?.pokemon_v2_type?.length || 0;

        return {
            uniquePokemonCount,
            duplicationPokemonCount,
            typesCount
        }
    }, [uniquePokemonsData, pokemonTypes, variationsPokemonData])


    return (
        <PokeStatsStyles>
            <div className={'stats-block'}>
                <div>Unique pokemons count: {stats.uniquePokemonCount}</div>
                <div>Pokemon duplications/variations count: {stats.duplicationPokemonCount}</div>
                <div>Total pokemons count: {stats.uniquePokemonCount + stats.duplicationPokemonCount}</div>
                <div>Types count: {stats.typesCount}</div>
            </div>
        </PokeStatsStyles>
    );
};

export default PokeStats;