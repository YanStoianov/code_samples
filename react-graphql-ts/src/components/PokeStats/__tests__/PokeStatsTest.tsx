import React from 'react'
import {render} from '@testing-library/react';
import 'jest-styled-components'
import PokeStats from "../PokeStats";
import {MockedProvider} from "@apollo/client/testing";


describe("PokeStatsTest", () => {
    beforeAll(() => {
        Object.defineProperty(window, "matchMedia", {
            writable: true,
            value: jest.fn().mockImplementation(query => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(),
                removeListener: jest.fn(),
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            }))
        });
    });

    test('snapshot', () => {
        const {asFragment} = render(
            <MockedProvider mocks={[]} addTypename={false}>
                <PokeStats pokemonTypes={undefined}/>
            </MockedProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    })
});