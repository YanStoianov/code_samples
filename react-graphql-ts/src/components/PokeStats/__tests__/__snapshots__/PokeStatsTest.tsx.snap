// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeStatsTest snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe bBEVAe"
  >
    <div
      class="stats-block"
    >
      <div>
        Unique pokemons count: 0
      </div>
      <div>
        Pokemon duplications/variations count: 0
      </div>
      <div>
        Total pokemons count: 0
      </div>
      <div>
        Types count: 0
      </div>
    </div>
  </div>
</DocumentFragment>
`;
